add errors tasks

sageichenko opened the pull request.

sterechenko committed:
add errors tasks

DenBal commented on errors.js:
```diff
@@ -0,0 +1,53 @@
+class ArgumentCountError extends CustomError {}
+
+const multiply = (number1, number2) => {
+    if ([...arguments].some((arg) => typeof arg !== 'number')) {
```
Стрелочные функции не имеют массива ````arguments`, поэтому тут будет ошибка:

` `````Uncaught ReferenceError: arguments is not defined`

sageichenko replied:
исправила этот позор..

DenBal commented on errors.js:
```diff
@@ -0,0 +1,53 @@
+        throw new ArgumentTypeError('Внимание! Один из аргументов метода не число');
+    }
+
+    if (arguments.length > 2) {
```
Тут будет ошибка из-за отсутствия массива

DenBal commented:
Task 1.

По логике работы функции multiply у тебя сначала идёт код, который проверяет тип переданных аргументов и, в случае ошибки, выбрасывает ошибку (строки 19-21), а потом идёт код проверки кол-ва переданных аргументов (строки 23-27), которые выводят в консоль warning. При этом если я укажу в качестве аргументов строки: '1', '2', '3', то я не увижу warning,  т.к. функция прекратит свою работу при выбросе ошибки. Это нужно поправить.

DenBal commented:
Task 1.

Что касается твоего вопроса про выброс предупреждения - ты все верно поняла, тут правильно использовать `console.warn`, выброс тут используется не в буквальном смысле.

DenBal commented:
Task 2.

На мой взгляд в методе try{} catch() {} не нужно ловить warning. Он у тебя и так описан внутри метода multiply, соотвественно при правильной логике метода multiply у тебя console.warn выведется. Можно вообще убрать его из try/catch.

DenBal commented:
Task 2.

Если я скопирую твой код из файла error.js перенесу его в браузерную консоль и вызову метод executeSomeHardMultiply(1, 2, 3). То результата никакого не будет, warning в консоли тоже не появится. Попробуй догадаться почему.

sterechenko committed:
Fix error after review

- fix use arguments in functions
- improve the order of checks
- remove catch ArgumentCountError

sageichenko commented:
> Task 2.
> 
> Если я скопирую твой код из файла error.js перенесу его в браузерную консоль и вызову метод executeSomeHardMultiply(1, 2, 3). То результата никакого не будет, warning в консоли тоже не появится. Попробуй догадаться почему.

потому что кто-то по привычке бездумно использует стрелочные функции) и не прокидывает ошибки дальше, если в catch не нашлось обработки для этого вида ошибки

DenBal commented on errors.js:
```diff
@@ -0,0 +1,47 @@
+    }
+}
+
+class ArgumentTypeError extends CustomError {}
```
Не до конца понимаю зачем тебе создавать CustomError и наследовать его для ArgumentTypeError. Обычный конструктор ошибки new Error() в качестве аргументов также принимает message.

Дока: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error

sageichenko replied:
CustomError изначально создавала, потому что планировала две кастомные ошибки, чтобы у них был общий родитель в этом модуле, т.е. я хотела отделить ошибки специфичные для этого модуля от всех остальных ошибок в коде. Название правда неудачное, ну и когда в итоге одна ошибка осталась, общий CustomError конечно совсем бессмысленно выглядит

DenBal replied:
Давай тогда его уберем и сделаем через Error

sageichenko replied:
Поправила

sterechenko committed:
Remove unneeded class CustomError
